runtime/src/main/kotlin/kotlin/native/internal/Cleaner.kt

```diff
@@ -0,0 +1,142 @@
+    private val cleanObj: (T) -> Unit,
+): Cleaner {
+
+    init {
+        // Make sure that Cleaner Worker is initialized.
+        // TODO: Is it needed?
+        CleanerWorker.worker
+    }
+
+    private val objHolder = StableRef.create(obj as Any)
+
+    @ExportForCppRuntime("Kotlin_CleanerImpl_clean")
+    private fun clean() {
+        // It's externally guaranteed that this is called only if cleanerWorker was
+        // not yet shut down.
+        val cleanPackage = Pair(cleanObj, objHolder).freeze()
+        CleanerWorker.worker.execute(TransferMode.SAFE, { cleanPackage }) { (cleanObj, objHolder) ->
+            try {
+                // TODO: Maybe if this fails with exception, it should be (optionally) reported.
+                @Suppress("UNCHECKED_CAST")
+                cleanObj(objHolder.get() as T)
+            } finally {
+                objHolder.dispose()
+            }
+        }
+    }
+}
+
+@SymbolName("Kotlin_Any_isShareable")
+external private fun Any?.isShareable(): Boolean
+
+@ExportForCompiler
+private fun <T> createCleanerImpl(argument: T, block: (T) -> Unit): Cleaner {
+    if (!argument.isShareable())
+        throw IllegalArgumentException("$argument must be shareable")
+
+    return CleanerImpl(argument, block.freeze())
+}
+
+/**
+ * Creates an object with a cleanup associated.
+ *
+ * After the resulting object ("cleaner") gets deallocated by memory manager,
+ * [block] is eventually called once with [argument].
+ *
+ * Example of usage:
+ * ```
+ * class ResourceWrapper {
+ *     private val resource = Resource()
+ *
+ *     private val cleaner = createCleaner(resource) { it.dispose() }
+ * }
+ * ```
+ *
+ * When `ResourceWrapper` becomes unused and gets deallocated, its `cleaner`
+ * is also deallocated, and the resource is disposed later.
+ *
+ * It is not specified which thread runs [block], as well as whether two or more
+ * blocks from different cleaners can be run in parallel.
+ *
+ * Note: if [argument] refers (directly or indirectly) the cleaner, then both
+ * might leak, and the [block] will not be called in this case.
+ * For example, the code below has a leak:
+ * ```
+ * class LeakingResourceWrapper {
+ *     private val resource = Resource()
+ *     private val cleaner = createCleaner(this) { it.resource.dispose() }
+ * }
+ * ```
+ * In this case cleaner's argument (`LeakingResourceWrapper`) can't be deallocated
+ * until cleaner's block is executed, which can happen only strictly after
+ * the cleaner is deallocated, which can't happen until `LeakingResourceWrapper`
+ * is deallocated. So the requirements on object deallocations are contradictory
+ * in this case, which can't be handled gracefully. The cleaner's block
+ * is not executed then, and cleaner and its argument might leak
+ * (depending on the implementation).
+ *
+ * [block] should not use `@ThreadLocal` globals, because it may
+ * be executed on a different thread.
+ *
+ * If [block] throws an exception, it gets silently ignored.
```

Comment:
I'm not sure it should work this way, but we haven't designed this case properly.
Let's just mention that it is not specified, and throwing an exception might do anything, from nothing to program termination?